factory.go

```diff
@@ -2,80 +2,59 @@ package goflow
-type Constructor func() (interface{}, error)
-
-// Annotation provides reference information about a component
-// to graph designers and operators
-type Annotation struct {
-	// Description tells what the component does
-	Description string
-	// Icon name in Font Awesome is used for visualization
-	Icon string
+// Factory registers components and creates their instances at run-time.
+// Not safe for concurrent use.
+type Factory struct {
+	registry map[string]registryEntry // map by the component name
 }
 
-// registryEntry contains runtime information about a component
+// registryEntry contains runtime information about a component.
 type registryEntry struct {
-	// Constructor is a function that creates a component instance.
-	// It is required for the factory to add components at run-time.
-	Constructor Constructor
-	// Run-time component description
-	Info ComponentInfo
-}
-
-// FactoryConfig sets up properties of a Factory
-type FactoryConfig struct {
-	RegistryCapacity uint
-}
-
-func defaultFactoryConfig() FactoryConfig {
-	return FactoryConfig{
-		RegistryCapacity: 64,
-	}
-}
-
-// Factory registers components and creates their instances at run-time
-type Factory struct {
-	registry map[string]registryEntry
+	constructor Constructor   // Function for creating component instances at run-time
+	info        ComponentInfo // Run-time component description
 }
 
-// NewFactory creates a new component Factory instance
-func NewFactory(config ...FactoryConfig) *Factory {
-	conf := defaultFactoryConfig()
-	if len(config) == 1 {
-		conf = config[0]
-	}
+// Constructor is used to create a component instance at run-time.
+type Constructor func() (interface{}, error)
 
+// NewFactory creates a new component Factory instance.
+func NewFactory() *Factory {
 	return &Factory{
-		registry: make(map[string]registryEntry, conf.RegistryCapacity),
+		registry: make(map[string]registryEntry),
 	}
 }
 
-// Register registers a component so that it can be instantiated at run-time
+// Register registers a component so that it can be instantiated at run-time.
 func (f *Factory) Register(componentName string, constructor Constructor) error {
 	if _, exists := f.registry[componentName]; exists {
 		return fmt.Errorf("registry error: component '%s' already registered", componentName)
 	}
 
 	f.registry[componentName] = registryEntry{
-		Constructor: constructor,
-		Info: ComponentInfo{
+		constructor: constructor,
+		info: ComponentInfo{
 			Name: componentName,
 		},
 	}
 
 	return nil
 }
 
-// Annotate adds human-readable documentation for a component to the runtime
+// Annotation provides reference information about a component to graph designers and operators.
+type Annotation struct {
+	Description string // Description of what the component does
+	Icon        string // Icon name in Font Awesome used for visualization
+}
+
+// Annotate adds human-readable documentation for a component to the runtime.
```

Comment:
Looking at the `Annotation`, both fields feel like something a component author would populate for their component to best describe the behavior. What is the rationale for annotating separately from the component registration?

Reply:
Annotation is entirely optional, while registration of a constructor in mandatory.

Annotations are typical for component libraries, where components are supposed to be reused outside of the original context. While for small homogenous applications developers probably know what the components are intended for since they wrote them.

Note: annotations are part of FBP-protocol support. FBP Protocol clients (e.g. Flowhub/NoFlo-UI) display annotations when browsing component libraries.